### DISCUSSION TITLE
Quic protocol support for invoking AWS Lambda

### DISCUSSION THREAD

**Discussion by bykka** [aws/aws-lambda-dotnet#1832] [Q&A]
Hi,

I have a problem invoking one lambda from another one. Both lambdas use .Net6 environment.
In the majority of cases, everything works fine but sometimes it fails with the next exception:

```
System.IO.FileLoadException: Could not load file or assembly 'System.Net.Quic, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The system cannot open the file.
 (0x80070004)
File name: 'System.Net.Quic, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'
   at System.Reflection.CustomAttribute._CreateCaObject(RuntimeModule pModule, RuntimeType type, IRuntimeMethodInfo pCtor, Byte** ppBlob, Byte* pEndBlob, Int32* pcNamedArgs)
   at System.Reflection.CustomAttribute.AddCustomAttributes(ListBuilder`1& attributes, RuntimeModule decoratedModule, Int32 decoratedMetadataToken, RuntimeType attributeFilterType, Boolean mustBeInheritable, ListBuilder`1 derivedAttributes)
```

Code example:
```csharp
var _lambdaClient = new AmazonLambdaClient();
 _lambdaClient.InvokeAsync(new()
  {
      FunctionName = "my-next-function",
      InvocationType = InvocationType.Event,
      Payload = JsonSerializer.Serialize(request)
  });
```

For me, it looks like the lambda client has got a request to upgrade the protocol to version HTTP3. But according to the [MS Documentation](https://learn.microsoft.com/en-us/dotnet/fundamentals/networking/quic/quic-overview) looks like this protocol became supported as a preview feature only in .Net7.

Does someone know how to fix it and why such an issue happens just time to time?

**Comment by normj** [comment]
The SDK and Lambda only support HTTP 1.1. There is nothing in the SDK that would attempt HTTP3, there are no AWS services that support HTTP3. I suspect the problem is coming from some where else in your application or something special on your environment that is trying to elevate the HTTP version, not sure what that would be though.

**Reply by bykka** [reply]
Thank you for your reply,

I execute my code in the lambda environment and do not control what happens there.
According to the log, it uses .net6.v44
`INIT_START Runtime Version: dotnet:6.v44`

And as I mentioned it is not a persistent issue - it happens only occasionally.

I would like to ask you what do you think, Is it possible that the `standard` AWS Lambda environment is not properly initiated?
Or is it a 100% issue on my side?

**Reply by normj** [reply]
I missed your code was running in Lambda. I was thinking maybe some IIS setting or policy was pushing for a higher version of HTTP but given you are running in Lambda that doesn't make sense. I can't see how anything in Lambda would try and upgrade the HTTP version. Sorry but I believe this is something going on in your side. The stacktrace's call stack does not call out as being part of the SDK call. Are you sure this exception is happening, occasionally, from the SDK's `InvokeAsync` call?

**Reply by bykka** [reply]
It looks like I have another issue that causes dll loading - `too many open files`.  But I do not work with files in the lambda - simply process dynamodb stream events and invoke another lambda in the end.
<img width="1451" alt="image" src="https://github.com/user-attachments/assets/b233a2e7-b9c9-4718-9800-2fda7289c41c">

**Reply by bykka** [reply]
I guess this article describes a similar case that I have got https://ttlnews.blogspot.com/2023/07/too-many-open-files-in-aws-lambda.html

In two worlds:  at night, when my lambda gets bursts of messages to process it opens to many HTTP connections (each new connection uses a file descriptor) for sending a message to another lambda.

**Reply by ashishdhingra** [reply]
@bykka Good morning. The "Too many open files" (IOException) error in C# AWS Lambda usually occurs when using Task.WhenAll on too many concurrent, un-awaited tasks, exceeding the 1024 Linux file descriptor limit. To fix this, ensure all async tasks are properly awaited, reuse HTTP clients outside the handler to prevent socket exhaustion, and implement throttling.

Looking at your code above, it appears that `_lambdaClient.InvokeAsync()` call is not awaited. Could you please use `await` keyword to the async call and check if issue gets resolved.

Thanks,
Ashish
